server/datastore/mysql/software.go

```diff
@@ -1721,3 +1721,307 @@ func (ds *Datastore) ListCVEs(ctx context.Context, maxAge time.Duration) ([]flee
+
+				WHEN hsi.install_script_exit_code IS NOT NULL AND
+					hsi.install_script_exit_code != 0 THEN ? -- failed
+
+				WHEN hsi.pre_install_query_output IS NOT NULL AND
+					hsi.pre_install_query_output = '' THEN ? -- failed
+
+				WHEN hsi.host_id IS NOT NULL THEN ? -- pending
+
+				ELSE NULL -- not installed from Fleet installer
+			END AS status,
+			si.id AS installer_id -- NULL if no Fleet installer
+		FROM
+			software_titles st
+		LEFT OUTER JOIN
+			software_installers si ON st.id = si.title_id
+		LEFT OUTER JOIN
+			host_software_installs hsi ON si.id = hsi.software_installer_id AND hsi.host_id = ?
+		WHERE
+			-- use the latest install only
+			( hsi.id IS NULL OR hsi.id = (
+				SELECT hsi2.id
+				FROM host_software_installs hsi2
+				WHERE hsi2.host_id = hsi.host_id AND hsi2.software_installer_id = hsi.software_installer_id
+				ORDER BY hsi2.created_at DESC
+				LIMIT 1 ) ) AND
+			-- software is installed on host
+			( EXISTS (
+				SELECT 1
+				FROM
+					host_software hs
+				INNER JOIN
+					software s ON hs.software_id = s.id
+				WHERE
+					hs.host_id = ? AND
+					s.title_id = st.id
+			) OR
+			-- or software install has been attempted on host
+			hsi.host_id IS NOT NULL )
+`
+
+	const stmtAvailable = `
+		SELECT
+			st.id,
+			st.name,
+			st.source,
+			si.filename as package_available_for_install,
+			NULL as last_install_installed_at,
+			NULL as last_install_install_uuid,
+			NULL as status,
+			si.id as installer_id
+		FROM
+			software_titles st
+		INNER JOIN
+			software_installers si ON st.id = si.title_id
+		WHERE
+			-- software is not installed on host, but is available in host's team
+			NOT EXISTS (
+				SELECT 1
+				FROM
+					host_software hs
+				INNER JOIN
+					software s ON hs.software_id = s.id
+				WHERE
+					hs.host_id = ? AND
+					s.title_id = st.id
+			) AND
+			-- sofware install has not been attempted on host
+			NOT EXISTS (
+				SELECT 1
+				FROM
+					host_software_installs hsi
+				WHERE
+					hsi.host_id = ? AND
+					hsi.software_installer_id = si.id
+			) AND
+			si.global_or_team_id = (SELECT COALESCE(h.team_id, 0) FROM hosts h WHERE h.id = ?)
+`
+
+	const selectColNames = `
+	SELECT
+		id,
+		name,
+		source,
+		package_available_for_install,
+		last_install_installed_at,
+		last_install_install_uuid,
+		status,
+		CASE
+			WHEN status = ? THEN 4 -- failed
+			WHEN status = ? THEN 3 -- pending
+			WHEN status = ? THEN 2 -- installed
+			WHEN installer_id IS NOT NULL THEN 1 -- installer exists, not installed
+			ELSE 0 -- no installer exists
+		END AS status_sort
+`
+
+	args := []any{
+		// for status_sort
+		fleet.SoftwareInstallerFailed,
+		fleet.SoftwareInstallerPending,
+		fleet.SoftwareInstallerInstalled,
+
+		// for status
+		fleet.SoftwareInstallerInstalled,
+		fleet.SoftwareInstallerFailed,
+		fleet.SoftwareInstallerInstalled,
+		fleet.SoftwareInstallerFailed,
+		fleet.SoftwareInstallerFailed,
+		fleet.SoftwareInstallerPending,
```

Comment:
just curious why you didn't go with `sqlx.Named`

Reply:
I probably should have! :D I'll add it to https://github.com/fleetdm/fleet/issues/18677 as a nice improvement.